merging a @ManyToOne association can call non-reactive code

Hi, 
I need to update 'erp_campus_id' in 'erp_academic_year_detail' table. The same code will works If I used normal hibernate. But In hibernate-reactive it throws an exception. I am using 1.0.0.CR9 version.

```java
AcademicYearDetaisDBO dbo = sessionFactory.withSession(session -> session.createQuery("select dbo from AcademicYearDetaisDBO dbo where dbo.id=386",AcademicYearDetaisDBO.class).getSingleResult()).await().indefinitely();

dbo.setRecordStatus('A');
System.out.println(dbo.getCampusId().getId());//for example here campus id is 11
CampusDBO campusDBO = new CampusDBO();
campusDBO.setId(5);
dbo.setCampusId(campusDBO); // need to update as 5

sessionFactory.withTransaction((session, transaction) -> session.merge(dbo)).await().indefinitely(); // here it will throw an exception
```



```java
@Getter
@Setter
@Entity
@Table(name = "erp_academic_year_detail")
public class AcademicYearDetaisDBO implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "erp_academic_year_detail_id")
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "erp_campus_id")
    private CampusDBO campusId;

    @Column(name = "record_status")
    private char recordStatus;

    @Column(name = "created_users_id")
    private Integer createdUsersId;

    @Column(name = "modified_users_id")
    private Integer modifiedUsersId;
}
```


```java
@Entity
@Getter
@Setter
@Table(name="erp_campus")
public class CampusDBO implements Serializable {
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        @Column(name="erp_campus_id")
        private  Integer id;

        @Column(name="campus_name")
        private String campusName;
}
```



Exception:


```
javax.persistence.PersistenceException: org.hibernate.HibernateException: java.util.concurrent.CompletionException: org.hibernate.reactive.event.impl.UnexpectedAccessToTheDatabase: Unexpected access to the database
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.reactive.session.impl.ReactiveExceptionConverter.convert(ReactiveExceptionConverter.java:31) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.lambda$fireMerge$28(ReactiveSessionImpl.java:842) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[na:1.8.0_301]
	at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:127) ~[asyncutil-0.1.0.jar:na]
	at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.lambda$unroll$0(AsyncTrampoline.java:123) ~[asyncutil-0.1.0.jar:na]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975) ~[na:1.8.0_301]
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:124) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:62) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:179) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:102) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.QueryResultBuilder.tryComplete(QueryResultBuilder.java:35) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.complete(Promise.java:66) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.handle(Promise.java:51) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.Promise.handle(Promise.java:29) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:124) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:62) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:179) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.TransactionImpl.lambda$wrap$0(TransactionImpl.java:72) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:124) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:62) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:179) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:279) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:98) ~[vertx-sql-client-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:97) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:240) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:130) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.lambda$handleMessage$9(NetSocketImpl.java:390) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:49) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:275) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:389) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155) ~[vertx-core-4.1.2.jar:4.1.2]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:154) ~[vertx-core-4.1.2.jar:4.1.2]



Caused by: java.util.concurrent.CompletionException: org.hibernate.reactive.event.impl.UnexpectedAccessToTheDatabase: Unexpected access to the database
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:673) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:683) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:2010) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:110) ~[na:1.8.0_301]
	at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.fetchAndCopyValues(DefaultReactiveMergeEventListener.java:484) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.lambda$null$8(DefaultReactiveMergeEventListener.java:348) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:110) ~[na:1.8.0_301]
	at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.lambda$entityIsDetached$10(DefaultReactiveMergeEventListener.java:348) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966) ~[na:1.8.0_301]
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940) ~[na:1.8.0_301]
	... 78 common frames omitted
Caused by: org.hibernate.reactive.event.impl.UnexpectedAccessToTheDatabase: Unexpected access to the database
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.onLoad(DefaultReactiveLoadEventListener.java:107) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:118) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.internal.SessionImpl.fireLoadNoChecks(SessionImpl.java:1215) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.internal.SessionImpl.internalLoad(SessionImpl.java:1080) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.EntityType.resolveIdentifier(EntityType.java:697) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.EntityType.resolve(EntityType.java:464) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.ManyToOneType.resolve(ManyToOneType.java:240) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.EntityType.resolve(EntityType.java:457) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.EntityType.replace(EntityType.java:358) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.type.TypeHelper.replace(TypeHelper.java:167) ~[hibernate-core-5.5.7.Final.jar:5.5.7.Final]
	at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.copyValues(DefaultReactiveMergeEventListener.java:493) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.lambda$fetchAndCopyValues$14(DefaultReactiveMergeEventListener.java:484) ~[hibernate-reactive-core-1.0.0.CR9.jar:1.0.0.CR9]
	at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:670) ~[na:1.8.0_301]


```

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/EntityTypes.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/EntityTypes.java
@@ -17,6 +17,7 @@ import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister;
 import org.hibernate.reactive.session.ReactiveQueryExecutor;
 import org.hibernate.type.EntityType;
+import org.hibernate.type.ForeignKeyDirection;
 import org.hibernate.type.OneToOneType;
 import org.hibernate.type.Type;
 import org.hibernate.type.TypeHelper;
@@ -168,7 +169,71 @@ public class EntityTypes {
                         owner,
                         copyCache
                 ).thenAccept( copy -> copied[i] = copy )
-        ).thenApply(v -> copied );
+        ).thenApply( v -> copied );
+    }
+
+    /**
+     * @see TypeHelper#replace(Object[], Object[], Type[], SharedSessionContractImplementor, Object, Map, ForeignKeyDirection)
+     */
+    public static CompletionStage<Object[]> replace(
+            final Object[] original,
+            final Object[] target,
+            final Type[] types,
+            final SessionImplementor session,
+            final Object owner,
+            final Map copyCache,
+            final ForeignKeyDirection foreignKeyDirection) {
+        Object[] copied = new Object[original.length];
+        for ( int i=0; i<types.length; i++ ) {
+            if ( original[i] == UNFETCHED_PROPERTY || original[i] == UNKNOWN ) {
+                copied[i] = target[i];
+            }
+            else {
+                if ( !(types[i] instanceof EntityType) ) {
+                    copied[i] = types[i].replace(
+                            original[i],
+                            target[i] == UNFETCHED_PROPERTY ? null : target[i],
+                            session,
+                            owner,
+                            copyCache,
+                            foreignKeyDirection
+                    );
+                }
+            }
+        }
+        return loop(0, types.length,
+                i -> original[i] != UNFETCHED_PROPERTY && original[i] != UNKNOWN
+                        && types[i] instanceof EntityType,
+                i -> replace(
+                        (EntityType) types[i],
+                        original[i],
+                        target[i] == UNFETCHED_PROPERTY ? null : target[i],
+                        session,
+                        owner,
+                        copyCache,
+                        foreignKeyDirection
+                ).thenAccept( copy -> copied[i] = copy )
+        ).thenApply( v -> copied );
+    }
+
+    /**
+     * @see org.hibernate.type.AbstractType#replace(Object, Object, SharedSessionContractImplementor, Object, Map, ForeignKeyDirection)
+     */
+    private static CompletionStage<Object> replace(
+            EntityType entityType,
+            Object original,
+            Object target,
+            SessionImplementor session,
+            Object owner,
+            Map copyCache,
+            ForeignKeyDirection foreignKeyDirection)
+            throws HibernateException {
+        boolean include = entityType.isAssociationType()
+                ? entityType.getForeignKeyDirection() == foreignKeyDirection
+                : ForeignKeyDirection.FROM_PARENT == foreignKeyDirection;
+        return include
+                ? replace( entityType, original, target, session, owner, copyCache )
+                : completedFuture(target);
     }
 
     /**
@@ -268,7 +333,7 @@ public class EntityTypes {
 //			// an entity type, in which case we need to resolve its identifier
 //			Type type = entityPersister.getPropertyType( entityType.uniqueKeyPropertyName );
 //			if ( type.isEntityType() ) {
-//				propertyValue = ( (EntityType) type ).getIdentifier( propertyValue, session );
+//				propertyValue = getIdentifier( (EntityType) type, propertyValue, session );
 //			}
 //
 //			return propertyValue;

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/EntityTypes.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/engine/impl/EntityTypes.java
@@ -1,0 +1,278 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.engine.impl;
+
+import org.hibernate.AssertionFailure;
+import org.hibernate.HibernateException;
+import org.hibernate.engine.spi.EntityKey;
+import org.hibernate.engine.spi.EntityUniqueKey;
+import org.hibernate.engine.spi.PersistenceContext;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister;
+import org.hibernate.reactive.session.ReactiveQueryExecutor;
+import org.hibernate.type.EntityType;
+import org.hibernate.type.OneToOneType;
+import org.hibernate.type.Type;
+import org.hibernate.type.TypeHelper;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+
+import static org.hibernate.bytecode.enhance.spi.LazyPropertyInitializer.UNFETCHED_PROPERTY;
+import static org.hibernate.property.access.internal.PropertyAccessStrategyBackRefImpl.UNKNOWN;
+import static org.hibernate.reactive.util.impl.CompletionStages.completedFuture;
+import static org.hibernate.reactive.util.impl.CompletionStages.loop;
+import static org.hibernate.reactive.util.impl.CompletionStages.nullFuture;
+
+public class EntityTypes {
+
+    /**
+     * Replacement for {@link EntityType#resolve(Object, SharedSessionContractImplementor, Object, Boolean)}
+     */
+    public static CompletionStage<Object> resolve(EntityType entityType, Object idOrUniqueKey, Object owner,
+                                           SharedSessionContractImplementor session) {
+        if ( idOrUniqueKey != null && !isNull( entityType, owner, session ) ) {
+            if ( entityType.isReferenceToPrimaryKey() ) {
+                return ((ReactiveQueryExecutor) session).reactiveInternalLoad(
+                        entityType.getAssociatedEntityName(),
+                        (Serializable) idOrUniqueKey,
+                        true,
+                        entityType.isNullable()
+                );
+            }
+            else {
+                return loadByUniqueKey( entityType, idOrUniqueKey, session );
+            }
+        }
+        else {
+            return null;
+        }
+    }
+
+    static boolean isNull(EntityType entityType, Object owner,
+                          SharedSessionContractImplementor session) {
+        if ( entityType instanceof OneToOneType) {
+            OneToOneType type = (OneToOneType) entityType;
+            String propertyName = type.getPropertyName();
+            if ( propertyName != null ) {
+                EntityPersister ownerPersister =
+                        session.getFactory().getMetamodel()
+                                .entityPersister( entityType.getAssociatedEntityName() );
+                Serializable id = session.getContextEntityIdentifier(owner);
+                EntityKey entityKey = session.generateEntityKey(id, ownerPersister);
+                return session.getPersistenceContextInternal().isPropertyNull( entityKey, propertyName);
+            }
+            else {
+                return false;
+            }
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * Load an instance by a unique key that is not the primary key.
+     *
+     * @param entityType The {@link EntityType} of the association
+     * @param key The unique key property value.
+     * @param session The originating session.
+     *
+     * @return The loaded entity
+     *
+     * @throws HibernateException generally indicates problems performing the load.
+     */
+    static CompletionStage<Object> loadByUniqueKey(
+            EntityType entityType,
+            Object key,
+            SharedSessionContractImplementor session) throws HibernateException {
+        SessionFactoryImplementor factory = session.getFactory();
+        String entityName = entityType.getAssociatedEntityName();
+        String uniqueKeyPropertyName = entityType.getRHSUniqueKeyPropertyName();
+
+        ReactiveEntityPersister persister =
+                (ReactiveEntityPersister) factory.getMetamodel().entityPersister( entityName );
+
+        //TODO: implement 2nd level caching?! natural id caching ?! proxies?!
+
+        EntityUniqueKey euk = new EntityUniqueKey(
+                entityName,
+                uniqueKeyPropertyName,
+                key,
+                entityType.getIdentifierOrUniqueKeyType( factory ),
+                persister.getEntityMode(),
+                factory
+        );
+
+        PersistenceContext persistenceContext = session.getPersistenceContextInternal();
+        Object result = persistenceContext.getEntity( euk );
+        if ( result != null ) {
+            return completedFuture( persistenceContext.proxyFor( result ) );
+        }
+        else {
+            return persister.reactiveLoadByUniqueKey( uniqueKeyPropertyName, key, session )
+                    .thenApply( loaded -> {
+                        // If the entity was not in the Persistence Context, but was found now,
+                        // add it to the Persistence Context
+                        if ( loaded != null ) {
+                            persistenceContext.addEntity(euk, loaded);
+                        }
+                        return loaded;
+                    } );
+
+        }
+    }
+
+    /**
+     * @see TypeHelper#replace(Object[], Object[], Type[], SharedSessionContractImplementor, Object, Map)
+     */
+    public static CompletionStage<Object[]> replace(
+            final Object[] original,
+            final Object[] target,
+            final Type[] types,
+            final SessionImplementor session,
+            final Object owner,
+            final Map copyCache) {
+        Object[] copied = new Object[original.length];
+        for ( int i=0; i<types.length; i++ ) {
+            if ( original[i] == UNFETCHED_PROPERTY || original[i] == UNKNOWN ) {
+                copied[i] = target[i];
+            }
+            else {
+                if ( !(types[i] instanceof EntityType) ) {
+                    copied[i] = types[i].replace(
+                            original[i],
+                            target[i] == UNFETCHED_PROPERTY ? null : target[i],
+                            session,
+                            owner,
+                            copyCache
+                    );
+                }
+            }
+        }
+        return loop(0, types.length,
+                i -> original[i] != UNFETCHED_PROPERTY && original[i] != UNKNOWN
+                        && types[i] instanceof EntityType,
+                i -> replace(
+                        (EntityType) types[i],
+                        original[i],
+                        target[i] == UNFETCHED_PROPERTY ? null : target[i],
+                        session,
+                        owner,
+                        copyCache
+                ).thenAccept( copy -> copied[i] = copy )
+        ).thenApply(v -> copied );
+    }
+
+    /**
+     * @see EntityType#replace(Object, Object, SharedSessionContractImplementor, Object, Map)
+     */
+    private static CompletionStage<Object> replace(
+            EntityType entityType,
+            Object original,
+            Object target,
+            SessionImplementor session,
+            Object owner,
+            Map copyCache) {
+        if ( original == null ) {
+            return nullFuture();
+        }
+        Object cached = copyCache.get( original );
+        if ( cached != null ) {
+            return completedFuture(cached);
+        }
+        else {
+            if ( original == target ) {
+                return completedFuture(target);
+            }
+            if ( session.getContextEntityIdentifier( original ) == null ) {
+                return ForeignKeys.isTransient( entityType.getAssociatedEntityName(), original, false, session )
+                        .thenCompose( isTransient -> {
+                            if ( isTransient ) {
+                                // original is transient; it is possible that original is a "managed" entity that has
+                                // not been made persistent yet, so check if copyCache contains original as a "managed" value
+                                // that corresponds with some "merge" value.
+                                if ( copyCache.containsValue( original ) ) {
+                                    return completedFuture(original);
+                                }
+                                else {
+                                    // the transient entity is not "managed"; add the merge/managed pair to copyCache
+                                    final Object copy = session.getEntityPersister( entityType.getAssociatedEntityName(), original )
+                                            .instantiate( null, session );
+                                    copyCache.put( original, copy );
+                                    return completedFuture(copy);
+                                }
+                            }
+                            else {
+                                return resolveIdOrUniqueKey( entityType, original, session, owner, copyCache );
+                            }
+                        } );
+            }
+            else {
+                return resolveIdOrUniqueKey( entityType, original, session, owner, copyCache );
+            }
+        }
+    }
+
+    private static CompletionStage<Object> resolveIdOrUniqueKey(
+            EntityType entityType,
+            Object original,
+            SessionImplementor session,
+            Object owner,
+            Map copyCache) {
+        return getIdentifier( entityType, original, session )
+                .thenCompose( id -> {
+                    if ( id == null ) {
+                        throw new AssertionFailure(
+                                "non-transient entity has a null id: "
+                                        + original.getClass().getName()
+                        );
+                    }
+                    Object idOrUniqueKey =
+                            entityType.getIdentifierOrUniqueKeyType( session.getFactory() )
+                                    .replace( id, null, session, owner, copyCache);
+                    return resolve( entityType, idOrUniqueKey, owner, session );
+                } );
+    }
+
+    /**
+     * see EntityType#getIdentifier(Object, SharedSessionContractImplementor)
+     */
+    private static CompletionStage<Serializable> getIdentifier(EntityType entityType, Object value, SessionImplementor session) {
+        if ( entityType.isReferenceToPrimaryKey()
+            /*|| entityType.uniqueKeyPropertyName == null*/ //TODO: expose this in core
+        ) {
+            return ForeignKeys.getEntityIdentifierIfNotUnsaved(
+                    entityType.getAssociatedEntityName(),
+                    value,
+                    session
+            ); //tolerates nulls
+        }
+        else if ( value == null ) {
+            return nullFuture();
+        }
+        else {
+            throw new UnsupportedOperationException("unique key properties not yet supported in merge()");
+            //TODO: expose stuff in core
+//			EntityPersister entityPersister = entityType.getAssociatedEntityPersister( session.getFactory() );
+//			Object propertyValue = entityPersister.getPropertyValue( value, entityType.uniqueKeyPropertyName );
+//			// We now have the value of the property-ref we reference.  However,
+//			// we need to dig a little deeper, as that property might also be
+//			// an entity type, in which case we need to resolve its identifier
+//			Type type = entityPersister.getPropertyType( entityType.uniqueKeyPropertyName );
+//			if ( type.isEntityType() ) {
+//				propertyValue = ( (EntityType) type ).getIdentifier( propertyValue, session );
+//			}
+//
+//			return propertyValue;
+        }
+    }
+
+}

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
@@ -215,7 +215,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		final EventSource source = event.getSession();
 		final EntityPersister persister = source.getEntityPersister( event.getEntityName(), entity );
 
-		copyCache.put( entity, entity, true );  //before cas‘cade!
+		copyCache.put( entity, entity, true );  //before cascade!
 
 		return cascadeOnMerge( source, persister, entity, copyCache )
 				.thenCompose( v -> fetchAndCopyValues( persister, entity, entity, source, copyCache ) )

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
@@ -253,12 +253,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		// copy created before we actually copy
 		//cascadeOnMerge(event, persister, entity, copyCache, Cascades.CASCADE_BEFORE_MERGE);
 		return super.cascadeBeforeSave( session, persister, entity, copyCache )
-				.thenAccept( v -> copyValues( persister, entity, copy, session, copyCache, FROM_PARENT ) )
+				.thenCompose( v -> copyValues( persister, entity, copy, session, copyCache, FROM_PARENT ) )
 				.thenCompose( v -> saveTransientEntity( copy, entityName, event.getRequestedId(), session, copyCache ) )
 				.thenCompose( v -> super.cascadeAfterSave( session, persister, entity, copyCache ) )
+				.thenCompose( v -> copyValues( persister, entity, copy, session, copyCache, TO_PARENT ) )
 				.thenAccept( v -> {
-					copyValues( persister, entity, copy, session, copyCache, TO_PARENT );
-
 					event.setResult(copy);
 
 					if (copy instanceof PersistentAttributeInterceptable) {
@@ -280,12 +279,9 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		//this bit is only *really* absolutely necessary for handling
 		//requestedId, but is also good if we merge multiple object
 		//graphs, since it helps ensure uniqueness
-		if ( requestedId == null ) {
-			return reactiveSaveWithGeneratedId( entity, entityName, copyCache, source, false );
-		}
-		else {
-			return reactiveSaveWithRequestedId( entity, requestedId, entityName, copyCache, source );
-		}
+		return requestedId == null
+				? reactiveSaveWithGeneratedId( entity, entityName, copyCache, source, false )
+				: reactiveSaveWithRequestedId( entity, requestedId, entityName, copyCache, source );
 	}
 
 	protected CompletionStage<Void> entityIsDetached(MergeEvent event, MergeContext copyCache) {
@@ -414,10 +410,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		// for enhanced entities, copy over the dirty attributes
 		if ( entity instanceof SelfDirtinessTracker && target instanceof SelfDirtinessTracker ) {
 			// clear, because setting the embedded attributes dirties them
-			( (SelfDirtinessTracker) target ).$$_hibernate_clearDirtyAttributes();
-
-			for ( String fieldName : ( (SelfDirtinessTracker) entity ).$$_hibernate_getDirtyAttributes() ) {
-				( (SelfDirtinessTracker) target ).$$_hibernate_trackChange( fieldName );
+			SelfDirtinessTracker entityTracker = (SelfDirtinessTracker) entity;
+			SelfDirtinessTracker targetTracker = (SelfDirtinessTracker) target;
+			targetTracker.$$_hibernate_clearDirtyAttributes();
+			for ( String fieldName : entityTracker.$$_hibernate_getDirtyAttributes() ) {
+				targetTracker.$$_hibernate_trackChange( fieldName );
 			}
 		}
 	}
@@ -504,7 +501,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		).thenAccept( copiedValues -> persister.setPropertyValues( target, copiedValues ) );
 	}
 
-	protected void copyValues(
+	protected CompletionStage<Void> copyValues(
 			final EntityPersister persister,
 			final Object entity,
 			final Object target,
@@ -512,13 +509,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 			final MergeContext copyCache,
 			final ForeignKeyDirection foreignKeyDirection) {
 
-		final Object[] copiedValues;
-
 		if ( foreignKeyDirection == TO_PARENT ) {
 			// this is the second pass through on a merge op, so here we limit the
 			// replacement to associations types (value types were already replaced
 			// during the first pass)
-			copiedValues = TypeHelper.replaceAssociations(
+			Object[] copiedValues = TypeHelper.replaceAssociations(
 					persister.getPropertyValues( entity ),
 					persister.getPropertyValues( target ),
 					persister.getPropertyTypes(),
@@ -527,9 +522,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 					copyCache,
 					foreignKeyDirection
 			);
+			persister.setPropertyValues( target, copiedValues );
+			return voidFuture();
 		}
 		else {
-			copiedValues = TypeHelper.replace(
+			return EntityTypes.replace(
 					persister.getPropertyValues( entity ),
 					persister.getPropertyValues( target ),
 					persister.getPropertyTypes(),
@@ -537,10 +534,8 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 					target,
 					copyCache,
 					foreignKeyDirection
-			);
+			).thenAccept( copiedValues -> persister.setPropertyValues( target, copiedValues ) );
 		}
-
-		persister.setPropertyValues( target, copiedValues );
 	}
 
 	/**
@@ -576,8 +571,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 	 * Cascade behavior is redefined by this subclass, disable superclass behavior
 	 */
 	@Override
-	protected CompletionStage<Void> cascadeAfterSave(EventSource source, EntityPersister persister, Object entity, MergeContext anything)
-			throws HibernateException {
+	protected CompletionStage<Void> cascadeAfterSave(EventSource source, EntityPersister persister, Object entity, MergeContext anything) {
 		return voidFuture();
 	}
 
@@ -585,8 +579,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 	 * Cascade behavior is redefined by this subclass, disable superclass behavior
 	 */
 	@Override
-	protected CompletionStage<Void> cascadeBeforeSave(EventSource source, EntityPersister persister, Object entity, MergeContext anything)
-			throws HibernateException {
+	protected CompletionStage<Void> cascadeBeforeSave(EventSource source, EntityPersister persister, Object entity, MergeContext anything) {
 		return voidFuture();
 	}
 

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveMergeEventListener.java
@@ -39,17 +39,21 @@ import org.hibernate.proxy.HibernateProxy;
 import org.hibernate.proxy.LazyInitializer;
 import org.hibernate.reactive.engine.impl.Cascade;
 import org.hibernate.reactive.engine.impl.CascadingAction;
+import org.hibernate.reactive.engine.impl.EntityTypes;
 import org.hibernate.reactive.event.ReactiveMergeEventListener;
 import org.hibernate.reactive.logging.impl.Log;
 import org.hibernate.reactive.logging.impl.LoggerFactory;
 import org.hibernate.reactive.session.ReactiveSession;
-import org.hibernate.reactive.util.impl.CompletionStages;
 import org.hibernate.service.ServiceRegistry;
 import org.hibernate.stat.spi.StatisticsImplementor;
 import org.hibernate.type.ForeignKeyDirection;
 import org.hibernate.type.TypeHelper;
 
+import static org.hibernate.event.internal.EntityState.DETACHED;
+import static org.hibernate.reactive.util.impl.CompletionStages.loop;
 import static org.hibernate.reactive.util.impl.CompletionStages.voidFuture;
+import static org.hibernate.type.ForeignKeyDirection.FROM_PARENT;
+import static org.hibernate.type.ForeignKeyDirection.TO_PARENT;
 
 /**
  * A reactific {@link org.hibernate.event.internal.DefaultMergeEventListener}.
@@ -84,11 +88,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		final EntityCopyObserver entityCopyObserver = createEntityCopyObserver( event.getSession().getFactory() );
 		final MergeContext mergeContext = new MergeContext( event.getSession(), entityCopyObserver );
 		return reactiveOnMerge(event, mergeContext)
-				.thenAccept(v -> entityCopyObserver.topLevelMergeComplete(event.getSession()))
-				.whenComplete((v, e) -> {
+				.thenAccept( v -> entityCopyObserver.topLevelMergeComplete( event.getSession() ) )
+				.whenComplete( (v, e) -> {
 					entityCopyObserver.clear();
 					mergeContext.clear();
-				});
+				} );
 	}
 
 	private EntityCopyObserver createEntityCopyObserver(SessionFactoryImplementor sessionFactory) {
@@ -172,7 +176,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 							// have an incoming entity instance which has a corresponding
 							// entry in the current persistence context, but registered
 							// under a different entity instance
-							entityState = EntityState.DETACHED;
+							entityState = DETACHED;
 						}
 					}
 				}
@@ -211,7 +215,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		final EventSource source = event.getSession();
 		final EntityPersister persister = source.getEntityPersister( event.getEntityName(), entity );
 
-		copyCache.put( entity, entity, true );  //before cascade!
+		copyCache.put( entity, entity, true );  //before cas‘cade!
 
 		return cascadeOnMerge( source, persister, entity, copyCache )
 				.thenCompose( v -> fetchAndCopyValues( persister, entity, entity, source, copyCache ) )
@@ -249,11 +253,11 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 		// copy created before we actually copy
 		//cascadeOnMerge(event, persister, entity, copyCache, Cascades.CASCADE_BEFORE_MERGE);
 		return super.cascadeBeforeSave( session, persister, entity, copyCache )
-				.thenAccept( v -> copyValues( persister, entity, copy, session, copyCache, ForeignKeyDirection.FROM_PARENT ) )
+				.thenAccept( v -> copyValues( persister, entity, copy, session, copyCache, FROM_PARENT ) )
 				.thenCompose( v -> saveTransientEntity( copy, entityName, event.getRequestedId(), session, copyCache ) )
 				.thenCompose( v -> super.cascadeAfterSave( session, persister, entity, copyCache ) )
 				.thenAccept( v -> {
-					copyValues(persister, entity, copy, session, copyCache, ForeignKeyDirection.TO_PARENT);
+					copyValues( persister, entity, copy, session, copyCache, TO_PARENT );
 
 					event.setResult(copy);
 
@@ -261,7 +265,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 						final PersistentAttributeInterceptable interceptable = (PersistentAttributeInterceptable) copy;
 						final PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
 						if (interceptor == null) {
-							persister.getBytecodeEnhancementMetadata().injectInterceptor(copy, id, session);
+							persister.getBytecodeEnhancementMetadata().injectInterceptor( copy, id, session );
 						}
 					}
 				});
@@ -323,7 +327,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 						// before cascade!
 						copyCache.put(entity, result, true);
 
-						Object target = unproxyManagedForDetachedMerging(entity, result, persister, source);
+						Object target = unproxyManagedForDetachedMerging( entity, result, persister, source );
 						if (target == entity) {
 							throw new AssertionFailure("entity was not detached");
 						}
@@ -334,7 +338,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 									entityName
 							);
 						}
-						else if ( isVersionChanged(entity, source, persister, target) ) {
+						else if ( isVersionChanged( entity, source, persister, target ) ) {
 							final StatisticsImplementor statistics = source.getFactory().getStatistics();
 							if (statistics.isStatisticsEnabled()) {
 								statistics.optimisticFailure(entityName);
@@ -476,30 +480,28 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 			// Cascade-merge mappings do not determine what needs to be fetched.
 			// The value only needs to be fetched if the incoming value (mergeState[i])
 			// is initialized, but its corresponding managed state is not initialized.
-			// Initialization must be done before copyValues executes.
-			stage = CompletionStages.loop(0, mergeState.length,
-										  i -> Hibernate.isInitialized( mergeState[i] ) && !Hibernate.isInitialized( managedState[i] ),
-										  i -> session.reactiveFetch( managedState[i], true ) );
+			// Initialization must be done before copyValues() executes.
+			stage = loop(0, mergeState.length,
+					i -> Hibernate.isInitialized( mergeState[i] ) && !Hibernate.isInitialized( managedState[i] ),
+					i -> session.reactiveFetch( managedState[i], true ) );
 		}
-		return stage.thenAccept( v -> copyValues( persister, entity, target, source, mergeContext ) );
+		return stage.thenCompose( v -> copyValues( persister, entity, target, source, mergeContext ) );
 	}
 
-	protected void copyValues(
+	protected CompletionStage<Void> copyValues(
 			final EntityPersister persister,
 			final Object entity,
 			final Object target,
 			final SessionImplementor source,
 			final MergeContext copyCache) {
-		final Object[] copiedValues = TypeHelper.replace(
+		return EntityTypes.replace(
 				persister.getPropertyValues( entity ),
 				persister.getPropertyValues( target ),
 				persister.getPropertyTypes(),
 				source,
 				target,
 				copyCache
-		);
-
-		persister.setPropertyValues( target, copiedValues );
+		).thenAccept( copiedValues -> persister.setPropertyValues( target, copiedValues ) );
 	}
 
 	protected void copyValues(
@@ -512,7 +514,7 @@ public class DefaultReactiveMergeEventListener extends AbstractReactiveSaveEvent
 
 		final Object[] copiedValues;
 
-		if ( foreignKeyDirection == ForeignKeyDirection.TO_PARENT ) {
+		if ( foreignKeyDirection == TO_PARENT ) {
 			// this is the second pass through on a merge op, so here we limit the
 			// replacement to associations types (value types were already replaced
 			// during the first pass)

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveResultSetProcessor.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/ReactiveResultSetProcessor.java
@@ -5,7 +5,6 @@
  */
 package org.hibernate.reactive.loader;
 
-import java.io.Serializable;
 import java.lang.invoke.MethodHandles;
 import java.sql.ResultSet;
 import java.sql.SQLException;
@@ -13,29 +12,20 @@ import java.util.List;
 import java.util.concurrent.CompletionStage;
 
 import org.hibernate.AssertionFailure;
-import org.hibernate.HibernateException;
 import org.hibernate.engine.internal.TwoPhaseLoad;
 import org.hibernate.engine.spi.EntityEntry;
-import org.hibernate.engine.spi.EntityKey;
-import org.hibernate.engine.spi.EntityUniqueKey;
 import org.hibernate.engine.spi.PersistenceContext;
 import org.hibernate.engine.spi.QueryParameters;
-import org.hibernate.engine.spi.SessionFactoryImplementor;
 import org.hibernate.engine.spi.SharedSessionContractImplementor;
 import org.hibernate.event.spi.PreLoadEvent;
 import org.hibernate.loader.plan.exec.query.spi.NamedParameterContext;
 import org.hibernate.loader.spi.AfterLoadAction;
-import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.reactive.engine.impl.EntityTypes;
 import org.hibernate.reactive.logging.impl.Log;
 import org.hibernate.reactive.logging.impl.LoggerFactory;
-import org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister;
-import org.hibernate.reactive.session.ReactiveQueryExecutor;
-import org.hibernate.reactive.util.impl.CompletionStages;
 import org.hibernate.transform.ResultTransformer;
-import org.hibernate.type.EntityType;
-import org.hibernate.type.OneToOneType;
 
-import static org.hibernate.reactive.util.impl.CompletionStages.completedFuture;
+import static org.hibernate.reactive.util.impl.CompletionStages.voidFuture;
 
 /**
  * An interface intended to unify how a ResultSet is processed by
@@ -72,16 +62,16 @@ public interface ReactiveResultSetProcessor {
 				session,
 				(entityType, value, source, owner, overridingEager)
 						-> entityType.isEager( overridingEager )
-								? resolve( entityType, (Serializable) value, owner, source )
+								? EntityTypes.resolve( entityType, value, owner, source )
 								: entityType.resolve( value, source, owner, overridingEager )
 		);
 
 		final Object[] hydratedState = entityEntry.getLoadedState();
-		CompletionStage<Void> loop = CompletionStages.voidFuture();
+		CompletionStage<Void> loop = voidFuture();
 		for ( int i = 0; i < hydratedState.length; i++ ) {
 			final Object o = hydratedState[i];
 			if ( o instanceof CompletionStage ) {
-				final CompletionStage<?> c = (CompletionStage) o;
+				final CompletionStage<?> c = (CompletionStage<?>) o;
 				final int currentIndex = i;
 				loop = loop.thenCompose( v -> c.thenAccept( initializedEntity -> hydratedState[ currentIndex ] = initializedEntity ) );
 			}
@@ -95,102 +85,5 @@ public interface ReactiveResultSetProcessor {
 						preLoadEvent
 				)
 		);
-	}
-
-	/**
-	 * Replacement for {@link EntityType#resolve(Object, SharedSessionContractImplementor, Object, Boolean)}
-	 */
-	default CompletionStage<Object> resolve(EntityType entityType, Serializable value, Object owner,
-											SharedSessionContractImplementor session) {
-		if ( value != null && !isNull( entityType, owner, session ) ) {
-			if ( entityType.isReferenceToPrimaryKey() ) {
-				return ((ReactiveQueryExecutor) session).reactiveInternalLoad(
-						entityType.getAssociatedEntityName(),
-						value,
-						true,
-						entityType.isNullable()
-				);
-			}
-			else {
-				return loadByUniqueKey( entityType, value, session );
-			}
-		}
-		else {
-			return null;
-		}
-	}
-
-	default boolean isNull(EntityType entityType, Object owner,
-						   SharedSessionContractImplementor session) {
-		if ( entityType instanceof OneToOneType ) {
-			OneToOneType type = (OneToOneType) entityType;
-			String propertyName = type.getPropertyName();
-			if ( propertyName != null ) {
-				EntityPersister ownerPersister =
-						session.getFactory().getMetamodel()
-								.entityPersister( entityType.getAssociatedEntityName() );
-				Serializable id = session.getContextEntityIdentifier(owner);
-				EntityKey entityKey = session.generateEntityKey(id, ownerPersister);
-				return session.getPersistenceContextInternal().isPropertyNull( entityKey, propertyName);
-			}
-			else {
-				return false;
-			}
-		}
-		else {
-			return false;
-		}
-	}
-
-	/**
-	 * Load an instance by a unique key that is not the primary key.
-	 *
-	 * @param entityType The {@link EntityType} of the association
-	 * @param key The unique key property value.
-	 * @param session The originating session.
-	 *
-	 * @return The loaded entity
-	 *
-	 * @throws HibernateException generally indicates problems performing the load.
-	 */
-	default CompletionStage<Object> loadByUniqueKey(
-			EntityType entityType,
-			Object key,
-			SharedSessionContractImplementor session) throws HibernateException {
-		SessionFactoryImplementor factory = session.getFactory();
-		String entityName = entityType.getAssociatedEntityName();
-		String uniqueKeyPropertyName = entityType.getRHSUniqueKeyPropertyName();
-
-		ReactiveEntityPersister persister =
-				(ReactiveEntityPersister) factory.getMetamodel().entityPersister( entityName );
-
-		//TODO: implement 2nd level caching?! natural id caching ?! proxies?!
-
-		EntityUniqueKey euk = new EntityUniqueKey(
-				entityName,
-				uniqueKeyPropertyName,
-				key,
-				entityType.getIdentifierOrUniqueKeyType( factory ),
-				persister.getEntityMode(),
-				factory
-		);
-
-		PersistenceContext persistenceContext = session.getPersistenceContextInternal();
-		Object result = persistenceContext.getEntity( euk );
-		if ( result != null ) {
-			return completedFuture( persistenceContext.proxyFor( result ) );
-		}
-		else {
-			return persister.reactiveLoadByUniqueKey( uniqueKeyPropertyName, key, session )
-					.thenApply( loaded -> {
-						// If the entity was not in the Persistence Context, but was found now,
-						// add it to the Persistence Context
-						if ( loaded != null ) {
-							persistenceContext.addEntity(euk, loaded);
-						}
-						return loaded;
-					} );
-
-		}
 	}
 }

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneMergeTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneMergeTest.java
@@ -1,0 +1,167 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import java.io.Serializable;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+import org.hibernate.cfg.Configuration;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import io.vertx.ext.unit.TestContext;
+
+public class ManyToOneMergeTest extends BaseReactiveTest {
+
+	@Override
+	protected Configuration constructConfiguration() {
+		Configuration configuration = super.constructConfiguration();
+		configuration.addAnnotatedClass( AcademicYearDetailsDBO.class );
+		configuration.addAnnotatedClass( CampusDBO.class );
+		return configuration;
+	}
+
+	@Before
+	public void populateDb(TestContext context) {
+		CampusDBO campusDBO2 = new CampusDBO();
+		campusDBO2.setId( 42 );
+		campusDBO2.setCampusName( "Kuchl" );
+
+		CampusDBO campusDBO = new CampusDBO();
+		campusDBO.setId( 66 );
+		campusDBO.setCampusName( "Qualunquelandia" );
+
+		AcademicYearDetailsDBO academicYearDetailsDBO = new AcademicYearDetailsDBO();
+		academicYearDetailsDBO.setId( 69 );
+		academicYearDetailsDBO.setCampusId( campusDBO );
+		academicYearDetailsDBO.setCreatedUsersId( 12 );
+		academicYearDetailsDBO.setRecordStatus( 'F' );
+		academicYearDetailsDBO.setModifiedUsersId( 66 );
+		test( context,
+			  getMutinySessionFactory().withTransaction( (session, transaction) -> session.persistAll(
+					  campusDBO, campusDBO2,
+					  academicYearDetailsDBO
+			  ) )
+		);
+	}
+
+	@Test
+	public void test(TestContext context) {
+		test( context, getMutinySessionFactory()
+				.withSession( session -> session
+						.createQuery( "select dbo from AcademicYearDetailsDBO dbo", AcademicYearDetailsDBO.class )
+						.getSingleResult() )
+				.chain( dbo -> {
+					dbo.setRecordStatus( 'A' );
+					System.out.println( dbo.getCampusId().getId() );//for example here campus id is 11
+					CampusDBO campusDBO = new CampusDBO();
+					campusDBO.setId( 5 );
+					dbo.setCampusId( campusDBO ); // need to update as 5
+					return getMutinySessionFactory()
+							.withTransaction( (session, transaction) -> {
+								dbo.setCampusId( session.getReference( CampusDBO.class, 42 ) );
+								return session.merge( dbo );
+							} );
+				} )
+		);
+	}
+
+	@Entity(name = "AcademicYearDetailsDBO")
+	@Table(name = "erp_academic_year_detail")
+	static class AcademicYearDetailsDBO implements Serializable {
+		@Id
+		@Column(name = "erp_academic_year_detail_id")
+		private Integer id;
+
+		// LAZY will work
+		@ManyToOne(fetch = FetchType.EAGER)
+		@JoinColumn(name = "erp_campus_id")
+		private CampusDBO campusId;
+
+		@Column(name = "record_status")
+		private char recordStatus;
+
+		@Column(name = "created_users_id")
+		private Integer createdUsersId;
+
+		@Column(name = "modified_users_id")
+		private Integer modifiedUsersId;
+
+		public Integer getId() {
+			return id;
+		}
+
+		public void setId(Integer id) {
+			this.id = id;
+		}
+
+		public CampusDBO getCampusId() {
+			return campusId;
+		}
+
+		public void setCampusId(CampusDBO campusId) {
+			this.campusId = campusId;
+		}
+
+		public char getRecordStatus() {
+			return recordStatus;
+		}
+
+		public void setRecordStatus(char recordStatus) {
+			this.recordStatus = recordStatus;
+		}
+
+		public Integer getCreatedUsersId() {
+			return createdUsersId;
+		}
+
+		public void setCreatedUsersId(Integer createdUsersId) {
+			this.createdUsersId = createdUsersId;
+		}
+
+		public Integer getModifiedUsersId() {
+			return modifiedUsersId;
+		}
+
+		public void setModifiedUsersId(Integer modifiedUsersId) {
+			this.modifiedUsersId = modifiedUsersId;
+		}
+	}
+
+	@Entity(name = "CampusDBO")
+	@Table(name = "erp_campus")
+	static class CampusDBO implements Serializable {
+		@Id
+		@Column(name = "erp_campus_id")
+		private Integer id;
+
+		@Column(name = "campus_name")
+		private String campusName;
+
+		public Integer getId() {
+			return id;
+		}
+
+		public void setId(Integer id) {
+			this.id = id;
+		}
+
+		public String getCampusName() {
+			return campusName;
+		}
+
+		public void setCampusName(String campusName) {
+			this.campusName = campusName;
+		}
+	}
+}

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneMergeTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/ManyToOneMergeTest.java
@@ -16,6 +16,7 @@ import javax.persistence.Table;
 
 import org.hibernate.cfg.Configuration;
 
+import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -48,10 +49,21 @@ public class ManyToOneMergeTest extends BaseReactiveTest {
 		academicYearDetailsDBO.setRecordStatus( 'F' );
 		academicYearDetailsDBO.setModifiedUsersId( 66 );
 		test( context,
-			  getMutinySessionFactory().withTransaction( (session, transaction) -> session.persistAll(
-					  campusDBO, campusDBO2,
-					  academicYearDetailsDBO
-			  ) )
+				getMutinySessionFactory()
+						.withTransaction( (session, transaction) -> session.persistAll(
+								campusDBO, campusDBO2,
+								academicYearDetailsDBO
+						) )
+		);
+	}
+
+	@After
+	public void cleanDb(TestContext context) {
+		test( context,
+				getMutinySessionFactory()
+						.withTransaction( (session, transaction) ->
+								session.createQuery("delete from AcademicYearDetailsDBO").executeUpdate()
+										.chain( v -> session.createQuery("delete from CampusDBO").executeUpdate() ) )
 		);
 	}
 
@@ -64,14 +76,40 @@ public class ManyToOneMergeTest extends BaseReactiveTest {
 				.chain( dbo -> {
 					dbo.setRecordStatus( 'A' );
 					System.out.println( dbo.getCampusId().getId() );//for example here campus id is 11
-					CampusDBO campusDBO = new CampusDBO();
-					campusDBO.setId( 5 );
-					dbo.setCampusId( campusDBO ); // need to update as 5
+//					CampusDBO campusDBO = new CampusDBO();
+//					campusDBO.setId( 5 );
+//					dbo.setCampusId( campusDBO ); // need to update as 5
 					return getMutinySessionFactory()
 							.withTransaction( (session, transaction) -> {
 								dbo.setCampusId( session.getReference( CampusDBO.class, 42 ) );
 								return session.merge( dbo );
 							} );
+				} )
+		);
+	}
+
+	@Test
+	public void testTransient(TestContext context) {
+		CampusDBO campusDBO = new CampusDBO();
+		campusDBO.setId( 77 );
+		campusDBO.setCampusName( "Qualunquelandia" );
+
+		AcademicYearDetailsDBO dbo = new AcademicYearDetailsDBO();
+		dbo.setId( 88 );
+		dbo.setCampusId( campusDBO );
+		dbo.setCreatedUsersId( 12 );
+		dbo.setRecordStatus( 'F' );
+		dbo.setModifiedUsersId( 66 );
+
+		test( context, getMutinySessionFactory()
+				.withSession( session -> {
+					dbo.setRecordStatus( 'A' );
+					System.out.println( dbo.getCampusId().getId() );//for example here campus id is 11
+//					CampusDBO cdbo = new CampusDBO();
+//					cdbo.setId( 5 );
+//					dbo.setCampusId( cdbo ); // need to update as 5
+					dbo.setCampusId( session.getReference( CampusDBO.class, 42 ) );
+					return session.merge( dbo );
 				} )
 		);
 	}

